junit-jupiter-api/src/main/java/org/junit/jupiter/api/DisplayNameGeneration.java

```diff
@@ -0,0 +1,168 @@
+ *
+ * @since 5.4
+ * @see DisplayName
+ * @see DisplayNameGenerator
+ */
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@API(status = EXPERIMENTAL, since = "5.4")
+public @interface DisplayNameGeneration {
+
+	Class<? extends DisplayNameGenerator> generator() default DisplayNameGenerator.class;
+
+	Style value() default Style.DEFAULT;
+
+	enum Style implements DisplayNameGenerator {
+		/**
+		 * Default display name generator.
+		 */
+		DEFAULT {
+			/**
+			 * TODO Javadoc
+			 */
+			@Override
+			public String generateDisplayNameForClass(Class<?> testClass) {
+				Preconditions.notNull(testClass, "Test class must not be null");
+				String name = testClass.getName();
+				int index = name.lastIndexOf('.');
+				return name.substring(index + 1);
+			}
+
+			/**
+			 * TODO Javadoc
+			 */
+			@Override
+			public String generateDisplayNameForNestedClass(Class<?> nestedClass) {
+				Preconditions.notNull(nestedClass, "Nested test class must not be null");
+				return nestedClass.getSimpleName();
+			}
+
+			/**
+			 * TODO Javadoc
+			 */
+			@Override
+			public String generateDisplayNameForMethod(Method testMethod) {
+				Preconditions.notNull(testMethod, "Test method must not be null");
+				return String.format("%s(%s)", testMethod.getName(),
+					ClassUtils.nullSafeToString(Class::getSimpleName, testMethod.getParameterTypes()));
+			}
+		},
+
+		/**
+		 * TODO Javadoc
+		 */
+		UNDERSCORE {
+			@Override
+			public String generateDisplayNameForClass(Class<?> testClass) {
+				return replaceUnderscore(DEFAULT.generateDisplayNameForClass(testClass));
+			}
+
+			@Override
+			public String generateDisplayNameForNestedClass(Class<?> nestedClass) {
+				return replaceUnderscore(DEFAULT.generateDisplayNameForNestedClass(nestedClass));
+			}
+
+			@Override
+			public String generateDisplayNameForMethod(Method testMethod) {
+				return replaceUnderscore(DEFAULT.generateDisplayNameForMethod(testMethod));
+			}
+
+			private String replaceUnderscore(String name) {
+				return name.replace('_', ' ');
+			}
+		},
+
+		/**
+		 * TODO Javadoc
+		 */
+		CAMEL_CASE {
+			private final Pattern camelCasePattern = Pattern.compile(String.format("%s|%s|%s",
+				"(?<=[A-Z])(?=[A-Z][a-z])", "(?<=[^A-Z])(?=[A-Z])", "(?<=[A-Za-z])(?=[^A-Za-z])"));
+
+			@Override
+			public String generateDisplayNameForClass(Class<?> testClass) {
+				return splitCamelCase(DEFAULT.generateDisplayNameForClass(testClass));
+			}
+
+			@Override
+			public String generateDisplayNameForNestedClass(Class<?> nestedClass) {
+				return splitCamelCase(DEFAULT.generateDisplayNameForNestedClass(nestedClass));
+			}
+
+			@Override
+			public String generateDisplayNameForMethod(Method testMethod) {
+				return splitCamelCase(DEFAULT.generateDisplayNameForMethod(testMethod));
+			}
+
+			private String splitCamelCase(String name) {
+				return camelCasePattern.matcher(name).replaceAll(" ");
+			}
+		},
+
+		/**
+		 * TODO Javadoc
+		 */
+		SENTENCES {
```

Comment:
Naming is difficult here.

I think it would help to prototype various examples with various levels of nesting to see what we are actually generating, and perhaps then a better name will "jump out at us".

Reply:
Good idea. Need a tons of templates for testing anyway, Morpheus says.